src/backend/distributed/transaction/backend_data.c

```diff
@@ -1278,3 +1290,202 @@ DecrementExternalClientBackendCounter(void)
 {
 	pg_atomic_sub_fetch_u32(&backendManagementShmemData->externalClientBackendCounter, 1);
 }
+
+
+/*
+ * GetEpochTimeMs returns the epoch value in milliseconds.
+ */
+uint64
+GetEpochTimeMs(void)
+{
+	struct timeval tp;
+
+	gettimeofday(&tp, NULL);
+
+	uint64 result = (uint64) (tp.tv_sec) * 1000;
+	result = result + (uint64) (tp.tv_usec) / 1000;
+	return result;
+}
+
+
+/*
+ * get_cluster_clock() is an UDF that returns a monotonically increasing logical
+ * clock. Clock guarantees to never go back in value after restarts, and makes
+ * best attempt to keep the value close to unix epoch time in milliseconds.
+ */
+Datum
+get_cluster_clock(PG_FUNCTION_ARGS)
+{
+	uint64 nextClusterClockValue = GetClusterClock();
+	PG_RETURN_UINT64(nextClusterClockValue);
+}
+
+
+uint64
+GetClusterClock(void)
+{
+	uint64 epochValue = GetEpochTimeMs();
+	uint64 nextClusterClockValue = 0;
+
+	SpinLockAcquire(&backendManagementShmemData->clockMutex);
+
+	/* Check if the clock is adjusted after the boot */
+	if (backendManagementShmemData->clusterClockValue == 0)
+	{
+		SpinLockRelease(&backendManagementShmemData->clockMutex);
+		ereport(ERROR, (errmsg("backend never adjusted the clock, please retry")));
```

Comment:
Or can it be the case that user have set citus.enable_global_clock GUC to false ?
If this might be the reason, then we might want to report sth like that;
```c
ereport(ERROR, (errmsg("cannot fetch cluster clock value, consider enabling it "
                       "by executing \"SET citus.enable_global_clock TO true\"")));
```

Reply:
That GUC plays role only for 2PC, please see **AdjustAndInitializeClusterClock()**, if the system is not using get_cluster_clock(), then we should have initialized it with Epoch time

Reply:
>>get_cluster_clock() can still be used even if that GUC is set to off, but it would then return the "local" clock
That's correct, as this is an UDF there is nothing that turns it on/off and when called it simply returns the local clock
>>citus.enable_global_clock 
If set to true, will simply orchestrate get_cluster_clock() (hidden from the user) across the nodes and stamps the transaction with the global clock (but it's currently not consumed)